29.md

```diff
@@ -0,0 +1,95 @@
+
+The entire metadata of the group (kinds `39000-3900x`) must only be signed by the relay main pubkey given by the `"pubkey"` field in the relay's [NIP-11](11.md) response. Other metadata events not signed by that public key must be ignored by clients. Upon opening a group, the chat client is supposed to fetch (or use locally cached values) these metadata events in order to render the group details, and then fetch the messages (`kind:9`).
+
+### Event definitions
+
+- *chat message* (`kind:9`)
+
+This is the basic unit of a message sent to a group.
+
+It must commit in the `g` tag to the relay hostname plus the subgroup path in order to prevent these events from being sent to other relays and groups out of context and cause confusion.
+
+`e` and `p` tags can also be used to signal an immediate reply to another message on the group or to call for the attention of some other group user.
+
+When inserting references -- in the body of the text -- to users, other messages (quotes) and to other Nostr events that outside of the group must be made following the [NIP-27](27.md) pattern (`nostr:nevent1...`).
+
+```js
+  "kind": 9,
+  "content": "hello my friends lovers of pizza",
+  "tags": [
+    ["g", "pizza.com/flavors"]
+  ]
+  ...
+```
+
+- *group metadata* (`kind:39000`) (optional)
+
+If this event does not exist, the group should be identified in the client UI by its identifier (i.e. "/flavors" or "pizza.com/flavors"). All tags are optional. Having the `"private"` tag means the group cannot be read and relays will use [NIP-42](42.md) `AUTH` messages to control who can read from it. The `"closed"` tag means the group can be read by anyone but only explicitly whitelisted pubkeys are allowed to post, again these enforcements happen at the relay level.
+
+```js
+{
+  "kind": 39000,
+  "content": "a nip-29 chat group for debating pizza flavors and other topics",
+  "tags": [
+    ["d", "/flavors"],
+    ["name", "Pizza Lovers"],
+    ["picture", "https://pizza.com/pizza.png"],
+    ["private"],
+    ["closed"]
+  ]
+  ...
+}
+```
+
+- *group admins* (`kind:39001`) (optional)
+
+Each admin gets a label that is only used for display purposes, and a list of permissions it has are listed afterwards. These permissions can inform client building UI, but ultimately are evaluated by the relay in order to become effective.
+
+The list of capabilities, as defined by this NIP, for now, is the following:
+
+- `add-user`
+- `edit-metadata`
+- `delete-message`
+- `ban-user`
+
+```js
+{
+  "kind": 39001,
+  "content": "list of admins for the pizza lovers group",
+  "tags": [
+    ["d", "/flavors"],
+    ["<pubkey1-as-hex>", "admin", "add-user", "edit-metadata", "delete-message", "ban-user"],
+    ["<pubkey2-as-hex>", "mod", "add-user", "delete-message"]
+  ]
+  ...
+}
+```
+
+- *moderation event* (`kind:39002`) (optional)
```

Comment:
Shouldn't this be a regular event `1000 <= n < 10000` (like `kind:1900`) to make it easier to implement? Cause if the relay already acted on the event (for example, already deleted a message), it would be very hard for the relay to undo if user wants to replace the event with a version without the "delete-message"  action.

Reply:
Yes, I got confused.